storage: differentiate between append-only and incremental storage-managed collections

aljoscha opened the pull request:
For now, only to shop around the concept. Please see notes/comments right in the code! 🙏 

### Motivation

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]

  * This PR adds a known-desirable feature.

    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

benesch commented on src/storage-controller/src/lib.rs:
```diff
@@ -797,17 +797,59 @@ where
+                            // desired state. No need to manually reset.
                         }
                         IntrospectionType::StorageSourceStatistics => {
+                            self.collection_manager.register_incremental_collection(id, read_handle_fn);
```
I'm not entirely sure that the stats should be modeled as incremental collections. I kind of think you might just want to toss the stats until the read-write promotion. Otherwise the cutover from the old version to the new version might cause the stats to regress, which will be quite surprising.

teskje commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -192,13 +346,377 @@ where
+                            }
+                            snapshot
+                        }
+                        Err(_) => panic!("read before since"),
```
This seems like something that can happen if a storage collection is created with a since != `[0]`. I think the storage controller will only bump the since on creation, not the upper. Can we handle this gracefully by first advancing the upper to the since?

teskje commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -192,13 +346,377 @@ where
+    );
+
+    (tx, handle.abort_on_drop(), shutdown_tx)
+}
```
I bet there is also some opportunity to deduplicate logic between the two task types.

aljoscha replied:
I'm sure there is! 🙈

teskje commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -192,13 +346,377 @@ where
+///
+/// NOTE: This implementation is a bit clunky, and could be optimized by not keeping
+/// all of desired in memory (see commend below). It is meant to showcase the
+/// general approach.
```
I think for unified compute introspection, doing this optimization would be pretty important. When discussing this feature in compute, we explicitly rules out keeping all the introspection data in memory to be able to retract it later, to avoid having envd memory usage scale with the number of replicas. It may be fine as long as we only have a few unified introspection collections, but I'd feel better to have this optimization in the initial implementation so we are not tempted to punt on it later.

teskje replied:
Discussed this offline: For the initial version of unified introspection the current way is fine. It actually simplifies the deletion of entries when a replica is dropped since we can just delete them from the in memory `desired` state, instead of having to read back the shard. As we start adding unified introspection collections we should start thinking about implementing the optimization.

teskje reviewed (commented):
Added a bunch of comments, the general approach looks good I think.

I'm still not entirely convinced that the added complexity is necessary though. I may be missing something, but for the 0dt use case it seems sufficient for the incremental write task to do a single retraction update when it comes out of read-only mode, to truncate the shard and fence out the older writer at the same time, then commence with directly appending the updates it receives like the appender task does.

aljoscha commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -51,7 +96,20 @@ pub struct CollectionManager<T>
 {
-    collections: Arc<Mutex<BTreeMap<GlobalId, (WriteChannel<T>, WriteTask, ShutdownSender)>>>,
+    // WIP: Name TBD! I thought about `managed_collections`, `ivm_collections`,
+    // `self_correcting_collections`.
```
I like it!

aljoscha commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -95,12 +154,63 @@ where
+                prev_task.is_finished(),
+                "should only spawn a new task if the previous is finished"
+            );
+        }
```
The comment above says that we double-check it. Maybe we want to be _extra_ sure. I just copied this code, which yes, is not a good excuse... 🤷‍♂️

aljoscha commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -95,12 +154,63 @@ where
+    where
+        R: FnMut() -> Pin<Box<dyn Future<Output = ReadHandle<SourceData, (), T, Diff>> + Send>>
+            + Send
+            + 'static,
```
Yeah that feels smart. Will do that if/when we figured out the bigger questions around this PR 👌

aljoscha commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -148,25 +258,66 @@ where
+    }
+
+    /// Updates the desired collection state of the collection identified by
+    /// `id`.
```
👌

aljoscha commented:
> I'm still not entirely convinced that the added complexity is necessary though. I may be missing something, but for the 0dt use case it seems sufficient for the incremental write task to do a single retraction update when it comes out of read-only mode, to truncate the shard and fence out the older writer at the same time, then commence with directly appending the updates it receives like the appender task does.

@teskje I did this for future proofing, to a degree. Even if we only wanted to do a single truncate/write when we come out of read mode, we would still have to have a retry loop around that, which has all of the logic that the current retry loop has. Because the old envd will continue trying to write (or could be) while we try and get in our retractions + fence. And we would then have two states: a) the initial state, where we try and get in our retractions/fenc, b) regular operations. Just having one state/retry loop felt simpler to me. And it's automatically resilient to future shenanigans where we have multiple writers of an introspection collection.

Interesting future scenario I wanted to be prepared for: multiple processes writing to a single introspection collection in parallel, but for example where each process is responsible only for a subset of the contents. The updates would be of the form:

```
Update(cluster_id, actual_update, ts, diff)
```

The writer for a `cluster_id` would filter out updates that are not for it's `cluster_id`, both when determining retractions and when writing updates. That way, it only has to keep the desired state of it's cluster in memory (if any). This has the benefit that we can share writing across different processes but still only have one collection, which makes it easy for consumers/readers. Maybe this is even already interesting for the compute introspection case?

teskje commented:
> Just having one state/retry loop felt simpler to me. And it's automatically resilient to future shenanigans where we have multiple writers of an introspection collection.

Fair enough. I'm not actually sure if the single-retraction-and-fencing approach would be less complicated than the self-correcting one. I figured not having to worry about consolidating often enough (but not too often) would simplify things. Also we may be able to save a bit of CPU and memory in read-write state by not having to do the diffing continually. But I'm not sure if either of these are significant. 

You could have made the complexity argument also for the self-correcting persist sink, but I think by now its clear that making it continually self-correcting was the right decision, based on how often this has simplified things elsewhere in the system.

> Interesting future scenario I wanted to be prepared for: multiple processes writing to a single introspection collection in parallel, but for example where each process is responsible only for a subset of the contents.

Hm, yes fencing doesn't really work in this scenario anymore, at least not if it's based on upper mismatches.

> This has the benefit that we can share writing across different processes but still only have one collection, which makes it easy for consumers/readers. Maybe this is even already interesting for the compute introspection case?

Maybe! We had considered having all replicas write their introspection data to persist directly, all to the same shard. The main wrinkle with this is how do you ensure things get retracted correctly when a replica drops or restarts? You needs something to write these updates down that's alive longer than the replica is. The current design has that thing be the controller and the controller writes introspection updates in a single thread, but that's not carved in stone.

teskje commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -192,13 +347,476 @@ where
+        self.to_write.extend(rows.iter().cloned());
+        self.desired.extend(rows.iter().cloned());
+
+        // TODO: Maybe don't do it every time?
```
You could consider using the `ConsolidatingVec` that's used by persist_sink. It only performs (amortized) O(log n) consolidation work per inserted update, at the cost of using up to twice the amount of memory of the fully consolidated vector. It's trading off CPU for memory which I'm not sure makes sense in this context.

teskje commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -192,13 +347,476 @@ where
+        // Note: it's very rare we should ever need to retry, and if we need to
+        // retry it should only take 1 or 2 attempts. We set `max_tries` to be
+        // high though because if we hit some edge case we want to try hard to
+        // commit the data.
```
Is that reasoning still the same now that we explicitly account for concurrent writers? I assume it is as long as we don't actually have them, but we might run into surprises when we start having them and forget about this code that limits the number of retries we are attempting.

aljoscha replied:
I think even with that it's good to limit the retries: we still don't shut down the write task, so the next time something in `desired` changes we would get a full set of retries again. That seems good, what do you think?

teskje replied:
Fine with me!

teskje commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -192,13 +347,476 @@ where
+        self.cmd_rx.close();
+
+        // Get as many waiting senders as possible.
+        'collect: while let Ok((_batch, sender)) = self.cmd_rx.try_recv() {
```
Nit: With the removal of the nested loop, this label isn't needed anymore.

teskje commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -192,13 +347,476 @@ where
+            .into_retry_stream();
+        let mut retries = Box::pin(retries);
+
+        'append_retry: loop {
```
Nit: With the removal of the nested loop, this label isn't needed anymore.

teskje commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -192,13 +347,476 @@ where
+                    let actual_upper = failed_ids[0]
+                        .current_upper
+                        .as_option()
+                        .expect("cannot write to empty collection")
```
Should we try to handle this gracefully? Or is there a reason why we can expect this to never happen?

teskje commented on src/storage-controller/src/lib.rs:
```diff
@@ -797,17 +797,59 @@ where
+                    // -> incremental collection
+                    //
+                    // Collection records things that happen over time, like a
+                    // history of something -> append-only collection.
```
This comment was more confusing to me than it helped, tbh. Isn't the distinction just "updates contain retractions" vs "updates don't contain retractions"?

teskje commented on src/storage-controller/src/lib.rs:
```diff
@@ -931,13 +995,20 @@ where
+                            // NOTE(aljoscha): We never remove from these
+                            // collections. Someone, at some point needs to
+                            // think about that!
+                            self.collection_manager.register_append_only_collection(id);
```
You could link the issue here, to ensure this comment doesn't become stale.

aljoscha replied:
Will do!

aljoscha commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -192,13 +347,476 @@ where
+        (tx, handle.abort_on_drop(), shutdown_tx)
+    }
+
+    async fn run(mut self) -> Result<(), ()> {
```
Yeah, again sorry about this unfinished piece. I'll change this to use `ControlFlow`, to make it more clear what's happening!

aljoscha commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -192,13 +347,476 @@ where
+        let batch_duration_ms = match self.id {
+            GlobalId::User(_) => {
+                Duration::from_millis(self.user_batch_duration_ms.load(Ordering::Relaxed))
+            }
```
For append-only collections this was added in order to rate-limit how many batches we write for webhook sources. They are also backed by this, and yes, I don't think it's a good idea anymore to have both storage-manage collections and webhooks be handled by these code paths. I think it conflates too much.

I will remove this flag from the incremental task.

aljoscha commented:
@teskje I rebased this and I think I addressed all your comments. I structured it such that you can follow along the `fixup!` commits while the base commit is as you reviewd it. Could you please take another look? 🙏

teskje commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -446,16 +447,26 @@ where
     }
 
-    async fn run(mut self) -> Result<(), ()> {
+    async fn run(mut self) -> ControlFlow<String> {
```
Nice, didn't know this existed! Makes things much clearer.

aljoscha replied:
Yeah, I only remembered it because persist uses it quite a bit: https://github.com/MaterializeInc/materialize/blob/afc85d2748e3489536c9ebc051b87fa96bf59b6c/src/persist-client/src/internal/state.rs#L956 😅

teskje reviewed (approved):
LGTM, thanks!

Given that this is a somewhat risky change and it's probably hard to feature-flag it in a meaningful way, it would be good to have some extra @MaterializeInc/testing scrutiny on this too!

nrainer-materialize commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -192,13 +340,477 @@ where
+                new_upper.clone(),
+            )];
+
+            let append_result = match self.write_handle.compare_and_append(request.clone()).await {
+                // We got a response!
+                Ok(append_result) => append_result,
+                // Failed to receive which means the worker shutdown.
+                Err(_recv_error) => {
+                    // Sender hung up, this seems fine and can happen when
+                    // shutting down.
+                    notify_listeners(responders, || {
+                        Err(StorageError::ShuttingDown("PersistMonotonicWriteWorker"))
+                    });
+
+                    // End the task since we can no longer send writes to persist.
+                    return ControlFlow::Break("sender hung up".to_string());
+                }
+            };
```
Not covered according to [coverage/428](https://buildkite.com/materialize/coverage/builds/428).

aljoscha replied:
Yeah, same as above. This code path is defensive, but the case that it protects against currently can't really happen. 🙈

nrainer-materialize commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -192,13 +340,477 @@ where
+
+                    // We've exhausted all of our retries, notify listeners and
+                    // break out of the retry loop so we can wait for more data.
+                    if retries.next().await.is_none() {
+                        notify_listeners(responders, || {
+                            Err(StorageError::InvalidUppers(failed_ids.clone()))
+                        });
+                        error!(
+                            "exhausted retries when appending to managed collection {failed_ids:?}"
+                        );
+                        break;
+                    }
```
Does not seem to be covered by tests according to [coverage/428](https://buildkite.com/materialize/coverage/builds/428).

aljoscha replied:
We can't currently trigger this code path because we don't have concurrent writers. We might have that in the future.

nrainer-materialize commented on src/storage-controller/src/lib.rs:
```diff
@@ -3199,11 +3237,58 @@ where
                 packer.extend(unpacked_row.into_iter());
-                (row_buf.clone(), -1)
+
+                let update = Update {
+                    row: row_buf.clone(),
+                    timestamp: expected_upper.clone(),
+                    diff: -1,
+                };
+
+                update
```
Does not seem to be covered by tests according to [coverage/428](https://buildkite.com/materialize/coverage/builds/428).

aljoscha replied:
This one is peculiar! The code was there before, I only added to it. I think it's only reached when a) there is enough updates in a status collection, and b) we do a restart of `envd`.

nrainer-materialize commented on src/storage-controller/src/lib.rs:
```diff
@@ -3199,11 +3237,58 @@ where
+            Ok(_) => {
+                // All good, yay!
+            }
+            Err(storage_err) => {
+                match storage_err {
+                    StorageError::InvalidUppers(failed_ids) => {
+                        assert_eq!(
+                            failed_ids.len(),
+                            1,
+                            "received errors for more than one collection"
+                        );
+                        assert_eq!(
+                            failed_ids[0].id, id,
+                            "received errors for a different collection"
+                        );
+
+                        // This is fine, it just means the upper moved because
+                        // of continual upper advancement or because seomeone
+                        // already appended some more retractions/updates.
+                        //
+                        // NOTE: We might want to attempt these partial
+                        // retractions on an interval, instead of only when
+                        // starting up!
+                        info!(%id, current_upper = ?failed_ids[0].current_upper, "failed to append partial truncation");
+                    }
+                    // Uh-oh, something else went wrong!
+                    other => {
+                        panic!("Unhandled error while appending to managed collection {id:?}: {other:?}")
+                    }
+                }
+            }
+        }
```
Does not seem to be covered by tests according to [coverage/428](https://buildkite.com/materialize/coverage/builds/428).

aljoscha replied:
Yeah, this also can't happen today because we can't have concurrent `envd` processes which are attempting to write. We might see this when we have real 0dt upgrades, where we temporarily have two `envd` processes alive.

So it's very much future-looking code.

nrainer-materialize reviewed (commented):
I annotated some spots that are not covered according to [coverage/428](https://buildkite.com/materialize/coverage/builds/428).

aljoscha commented on src/storage-controller/src/collection_mgmt.rs:
```diff
@@ -192,13 +340,477 @@ where
+        ControlFlow::Continue(())
+    }
+
+    fn handle_shutdown(&mut self) {
+        let mut senders = Vec::new();
+
+        // Prevent new messages from being sent.
+        self.cmd_rx.close();
+
+        // Get as many waiting senders as possible.
+        while let Ok((_batch, sender)) = self.cmd_rx.try_recv() {
+            senders.push(sender);
+
+            // Note: because we're shutting down the sending side of `rx` is no
+            // longer accessible, and thus we should no longer receive new
+            // requests. We add this check just as an extra guard.
+            if senders.len() > CHANNEL_CAPACITY {
+                // There's not a correctness issue if we receive new requests,
+                // just unexpected behavior.
+                tracing::error!("Write task channel should not be receiving new requests");
+                break;
+            }
+        }
+
+        // Notify them that this collection is closed.
+        //
+        // Note: if a task is shutting down, that indicates the source has been
+        // dropped, at which point the identifier is invalid. Returning this
+        // error provides a better user experience.
+        notify_listeners(senders, || Err(StorageError::IdentifierInvalid(self.id)));
+    }
```
Yes, I should have expected that! This code path is only taken when removing a collection that is manged by this. But this code is only used for builtin storage-managed collections (the introspection collections), and we normally never remove them, or can't even do that.

We _might_ add code in the future that does drop these collections, in which case this code path will come in handy, but it might rot! Would you prefer I remove this, and add something akin to `unreachable!()`, or leave it in?

nrainer-materialize replied:
Something like `unreachable!()` sounds good

aljoscha commented:
Thanks everyone for the review and patience with this one! 🙏
